inseq/attr/feat/attention_attribution.py

```diff
@@ -0,0 +1,128 @@
+# Copyright 2021 The Inseq Team. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+""" Attention-based feature attribution methods. """
+
+from typing import Any, Dict, Union
+
+import logging
+
+from ...data import FeatureAttributionStepOutput
+from ...utils import Registry
+from ...utils.typing import ModelIdentifier
+from ..attribution_decorators import set_hook, unset_hook
+from .feature_attribution import FeatureAttribution
+from .ops import AggregatedAttention, LastLayerAttention
+
+
+logger = logging.getLogger(__name__)
+
+
+class AttentionAtribution(FeatureAttribution, Registry):
```

Comment:
Override `format_attribute_args` here. We don't care to preserve the exact format input of gradient methods. A good set of input params for `attribute` would be:
- `batch`: pass the full batch to the attribution method
- If we pass the full batch, we don't need `additional_forward_args`
- Attention methods do not rely on the output, so `target_ids` are superfluous

Reply:
Does it make sense to replace those arguments fully here? Because I thought the format might also be used for custom attribution functions later in the pipeline?

Reply:
It makes sense to change it, since it is entirely related to the individual attribution method. Custom attributed functions have different parameters which are independent from the attribution method ones, which you can find described in the [`register_step_score` docstring](https://github.com/inseq-team/inseq/blob/9e88a9ebf3788043ec6cc0c6fa334f594a29b154/inseq/attr/feat/attribution_utils.py#L191). Moreover, the attention methods should check whether `attribution_model.default_attributed_fn_id` is different from the default, and if yes warn the user that attention scores are independent of the specific generated target and the attributed function at every specific step.